### ISSUE TITLE
Block-level data support

### ISSUE THREAD

**Issue by everyonesdesign** [springload/draftjs_exporter#39]
I began to use draft-js and I discovered a feature called 'block-level data'. Via [Modifier#setBlockData](https://facebook.github.io/draft-js/docs/api-reference-modifier.html#setblockdata) you can set additional metadata for a specific block.

In that case the generated JSON looks like this:
```
{  
   "entityMap": {},
   "blocks": [  
      ...
      {  
         "key": "a4dpo",
         "text": "asdfasdfsadfatsetsd",
         "type": "direct-speech",
         "depth": 0,
         "inlineStyleRanges": [],
         "entityRanges": [],
         "data": {  
            "name": "test"
         }
      },
      ...
   ]
}
```
Is there any way to implement a block rendering depending on these attributes values?

**Comment by thibaudcolas** [comment]
Hey @everyonesdesign,

This is not supported at the moment. I would love to support this use case but I am not using this feature of Draft.js in my editor implementation, so I can't commit to implementing it myself any time soon. That said, I'm very happy to discuss and review a PR if you are keen to implement it yourself.

Here is what it could look like, following the API of entity decorators:

```python
class LI:
    def render(self, props):
        # Not sure whether data is the right name for this key, or "metadata" or "block_data".
        # Ideally this should be aligned on the React props.
        data = props.get('data')
        return DOM.create_element('li')

config = {
    'entity_decorators': {
        ENTITY_TYPES.LINK: Link(),
    },
    'block_map': dict(BLOCK_MAP, **{
        BLOCK_TYPES.UNORDERED_LIST_ITEM: {
            'element': LI(),
            'wrapper': ['ul', {'className': 'bullet-list'}],
        },
    }),
}

exporter = HTML(config)
```

To implement this, I think it could be enough to change https://github.com/springload/draftjs_exporter/blob/master/draftjs_exporter/wrapper_state.py#L131 to pass the block's `data` as a prop. You could try it yourself by changing this one line to:

```python
elt = DOM.create_element(options.element[0], block.get('data'))
```

Then the next step would be figuring out how this works with the different block map formats documented here: https://github.com/springload/draftjs_exporter/blob/master/draftjs_exporter/wrapper_state.py#L23.

Finally I would probably want to make sure that `wrapper` supports this too, but that doesn't have to happen at the same time.

Now the reason I'm not using this / supporting this at the moment is because I haven't found good enough documentation on block-level metadata (what it's meant for) to risk relying on it. Do you have any good resources on the topic?

**Comment by everyonesdesign** [comment]
@thibaudcolas

Thank you very much for your detailed explanation.

I'll check out the changes you suggested.

When I implemented a feature based on block-level data, I found [this issue](https://github.com/facebook/draft-js/issues/129), [this request](https://github.com/facebook/draft-js/pull/216), aforementioned [`Modifier#setBlockData`](https://facebook.github.io/draft-js/docs/api-reference-modifier.html#setblockdata) method and [ContentBlock#getData](https://facebook.github.io/draft-js/docs/api-reference-content-block.html#getdata) method. 

I also checked Draft source code to make sure it works the way I expect it to.

I have to admit that the documentation of Draft is really shallow at the time and this feature is poorly documented.

**Comment by thibaudcolas** [comment]
Just merged #40 which introduces support for this. Thank you for starting this @everyonesdesign! 

I will be releasing this as part of 0.8.0 very soon, which introduces a lot of breaking changes, but hopefully should be the last release before 1.0.0 (and then the API/config will be more stable).

The `example.py` file now demonstrates how to use this to render blocks based on their depth and/or data (https://github.com/springload/draftjs_exporter/blob/5a92deea36a17615849431cb7e202c68d10c87ac/example.py#L17), and there should be more formal documentation available soon.
